### ISSUE TITLE
GEOSException with hash on POINT EMPTY

### ISSUE THREAD

**Issue by mwtoews** [pygeos/pygeos#196]
PyGEOS now has two good features:
 * #102 hash support
 * #179 `POINT EMPTY` supported by `to_wkb`

It would be expected that `POINT EMPTY` could be hashed, however GEOSException is raised:
```python
from pygeos import Geometry, to_wkb

g = Geometry('POINT EMPTY')

to_wkb(g)  # now works!

hash(g)
---------------------------------------------------------------------------
GEOSException                             Traceback (most recent call last)
<ipython-input-34-ace70842997c> in <module>
----> 1 hash(g)

GEOSException: IllegalArgumentException: Empty Points cannot be represented in WKB
```

**Comment by mwtoews** [comment]
Looking at `point_nan` from 8d1b66d47083054e9c9ff23a0c9787a9b54a44a3, this is not an empty point:
```python
import pygeos
import numpy as np

point_nan = pygeos.points(np.nan, np.nan)
point_empty = pygeos.Geometry('point empty')

pygeos.is_empty(point_nan)  # False
pygeos.is_empty(point_empty)  # True

repr(point_nan)  # <pygeos.Geometry POINT (nan nan)>
repr(point_empty)  # <pygeos.Geometry POINT EMPTY>

hash(point_nan)  # 8484330208244771434
# hash(point_empty)  # crash

pygeos.to_wkb(point_nan) == pygeos.to_wkb(point_empty) # True
```
This behaviour (apart from the hash crash) isn't wrong, and is similar to PostGIS (via lwgeom):
```sql
SELECT ST_AsEWKT(point_nan), ST_AsEWKT(point_empty),
  ST_IsEmpty(point_nan), ST_IsEmpty(point_empty),
  ST_AsEWKT(point_nan) = ST_AsEWKT(point_empty) AS wkt_eq,
  ST_AsEWKB(point_nan) = ST_AsEWKB(point_empty) AS wkb_eq
FROM (
  SELECT ST_Point('nan'::float, 'nan'::float) AS point_nan,
                      'POINT EMPTY'::geometry AS point_empty
) AS f;
-- -[ RECORD 1 ]--------------
-- st_asewkt  | POINT(nan nan)
-- st_asewkt  | POINT EMPTY
-- st_isempty | f
-- st_isempty | t
-- wkt_eq     | f
-- wkb_eq     | t
```
So these two points have different WKT but same WKB.

**Comment by mwtoews** [comment]
and while I'm on it, this is also a bit of a surprise:
```python
point_nan_from_wkt = pygeos.Geometry('POINT(nan nan)')
pygeos.is_empty(point_nan_from_wkt)  # True
repr(point_nan_from_wkt)  # <pygeos.Geometry POINT EMPTY>
```
(same from shapely too).

**Comment by jorisvandenbossche** [comment]
@mwtoews Thanks for the catch! We patched to_wkb to deal with empty points, but didn't used those patched in the WKB conversion code in the hash function. 
Now, I think https://github.com/pygeos/pygeos/pull/190 is already trying to deduplicate this code, and probably fixes the reported issue.

> Looking at point_nan from 8d1b66d, this is not an empty point:

Well, that's the consequence of the choice to use Point(NaN, NaN) to signify empty points in WKB, nothing to do about that I think .. 
(and personally I think this is fine, since Point(NaN, NaN) should be even more rare than empty points?)

**Comment by jorisvandenbossche** [comment]
Regarding:

```
In [9]: pygeos.from_wkt("POINT (nan nan)")
Out[9]: <pygeos.Geometry POINT EMPTY>
```

I suppose that is behaviour of GEOS? (as we don't do any patching here for WKT conversions) 
So in a sense, GEOS is *somewhat* trying to be consistent in WKT and WKB (at least for reading, but not for writing ..)

**Comment by mwtoews** [comment]
> I suppose that is behaviour of GEOS?

Yes, I consider this an inconsistency that should be addressed there; see [GEOS trac#1049](https://trac.osgeo.org/geos/ticket/1049).

**Comment by caspervdw** [comment]
Thanks for the report. I confirmed that this segfault is solved in master (see  #196).

The other issue about conversions between POINT (nan nan) and POINT EMPTY are now described in the docs (https://pygeos.readthedocs.io/en/latest/io.html#pygeos.io.to_wkb)
